SMP support for hardware RISC-V boards, including Microchip's PolarFire SoC Icicle Kit

con-pax opened the pull request:
SMP support for RISC-V implementations based on the experience of implementing SMP on The PolarFire Soc Icicle Kit. The changes provide support for heterogeneous core complexes by:
- Configuration option to choose the boot hart,
- Allowing selections of CPUs for MP and SMP applications using Devicetree overlays.

The changes in this PR where first published in the pr #46188
Some of the most prominent feedback was that the configuration options where too general, as such, changes were applied to the RISC-V layer.
This PR depends upon PR #54143

npitre commented:
Many comments:

- Please make a separate PR for the `z_smp_cpu_mobile()` changes.
  This is not RISC-V specific and deserves its own PR.

- Then you should introduce `CONFIG_RV_BASE_CPU` in a commit of its own.
  That is on itself pretty uncontrovertial although it has implications
  in may different places and should stand on its own.

- Please avoid construct like this:

```
+#if CONFIG_RV_BASE_CPU > 0
+               hartid -= CONFIG_RV_BASE_CPU;
+#endif
```

  You should dispense with `#if` and let the compiler optimize the
  subtraction-by-0 away.

- You have this:

```
+#ifdef CONFIG_RV_BASE_CPU
 #define GET_CURRENT_CPU(dst, tmp) \
        csrr tmp, mhartid                               ;\
+       addi tmp, tmp, -CONFIG_RV_BASE_CPU      ;\
        la dst, _kernel + ___kernel_t_cpus_OFFSET       ;\
        shiftmul_add dst, tmp, ___cpu_t_SIZEOF
+#else
+#define GET_CURRENT_CPU(dst, tmp) \
+       csrr tmp, mhartid                               ;\
+       la dst, _kernel + ___kernel_t_cpus_OFFSET       ;\
+       shiftmul_add dst, tmp, ___cpu_t_SIZEOF
+#endif
```

You could consider this alternative:

```
 #define GET_CURRENT_CPU(dst, tmp) \
        csrr tmp, mhartid                               ;\
-       la dst, _kernel + ___kernel_t_cpus_OFFSET       ;\
+       la dst, _kernel + ___kernel_t_cpus_OFFSET - (CONFIG_RV_BASE_CPU * ___cpu_t_SIZEOF) ;\
        shiftmul_add dst, tmp, ___cpu_t_SIZEOF
```

- The mcounteren and "monitor heart" special boot treatments should be
  relegated to a file specific to your own SOC and not stuffed in the
  generic code. See `zephyr/soc/riscv-privilege/telink_b91/start.S` or
  `zephyr/soc/./riscv-privilege/neorv32/reset.S` for examples.

- If `CONFIG_SUPERVISOR_MODE_CAPABLE` refer to generic issues then related
  changes should have a commit or their own. If specific to your SOC then
  previous comment should apply. Al;so this Kconfig symbol should
  be more RISC-V specific in its name. Something like
  `CONFIG_RISCV_S_MODE_CAPABLE` perhaps.
i

- You have:

```
+K_KERNEL_STACK_ARRAY_DEFINE(_non_z_stacks,
+                            1,
+                            CONFIG_ISR_STACK_SIZE);
```

  If this is a non Zephyr stack as its name implies, and to be used by
  non Zephyr code, then this has no business here. You should allocate
  stack space alongside the code that will be using it. Furthermore the
  `K_KERNEL_STACK_ARRAY_DEFINE()` allocates more than just the provided
  size so to be compatible with PMP stack guard which non-Zephyr code is
  not likely to benefit from.

- You are reintroducing the memory corruption bug that was fixed in
  PR #50459.

- And finally, given that all the secondary boot gating changes should
  be isolated to a commit of their own at this point, could you elaborate
  in the commit log why those changes are needed and what problem they
  solve? I think this part deserves more scrutiny and not just be some
  carried-over code as the previous comment might imply.

con-pax commented:
@npitre Thank you for the comments! we will get on them right away.

petermcs-01 commented:
> * Please avoid construct like this:
> 
> ```
> +#if CONFIG_RV_BASE_CPU > 0
> +               hartid -= CONFIG_RV_BASE_CPU;
> +#endif
> ```
> 
> You should dispense with `#if` and let the compiler optimize the subtraction-by-0 away.
> 
> * You have this:
> 
> ```
> +#ifdef CONFIG_RV_BASE_CPU
>  #define GET_CURRENT_CPU(dst, tmp) \
>         csrr tmp, mhartid                               ;\
> +       addi tmp, tmp, -CONFIG_RV_BASE_CPU      ;\
>         la dst, _kernel + ___kernel_t_cpus_OFFSET       ;\
>         shiftmul_add dst, tmp, ___cpu_t_SIZEOF
> +#else
> +#define GET_CURRENT_CPU(dst, tmp) \
> +       csrr tmp, mhartid                               ;\
> +       la dst, _kernel + ___kernel_t_cpus_OFFSET       ;\
> +       shiftmul_add dst, tmp, ___cpu_t_SIZEOF
> +#endif
> ```
> 
> You could consider this alternative:
> 
> ```
>  #define GET_CURRENT_CPU(dst, tmp) \
>         csrr tmp, mhartid                               ;\
> -       la dst, _kernel + ___kernel_t_cpus_OFFSET       ;\
> +       la dst, _kernel + ___kernel_t_cpus_OFFSET - (CONFIG_RV_BASE_CPU * ___cpu_t_SIZEOF) ;\
>         shiftmul_add dst, tmp, ___cpu_t_SIZEOF
> ```
These will only work if the macro is defined and will produce compiler errors otherwise hence the #if/#ifdef

petermcs-01 commented:
> * The mcounteren and "monitor heart" special boot treatments should be
>   relegated to a file specific to your own SOC and not stuffed in the
>   generic code. 
The mcounteren register needs to be set on real hardware otherwise the registers won't be accessible from user mode... - I don't know if QEMU honors this but i suspect not if the test was passing previously. 

In terms of the monitor heart, there are at least two implementations out there that support this sort of setup and more on the horizon so I think it is more generally useful.

petermcs-01 commented:
> * And finally, given that all the secondary boot gating changes should
>   be isolated to a commit of their own at this point, could you elaborate
>   in the commit log why those changes are needed and what problem they
>   solve? I think this part deserves more scrutiny and not just be some
>   carried-over code as the previous comment might imply.
The improvements to the secondary CPU bring up are not carried over code but a genuine part of this commit. My experience both on bare metal drivers and Zephyr development with multiple heart systems has shown that when debugging in particular, the use of WFI alone is not sufficient control the orderly bring up of the hearts as the debugger tends to trigger escapes from WFI which causes synchronization issues without some extra state management at the software level.

petermcs-01 commented:
> * You have:
> 
> ```
> +K_KERNEL_STACK_ARRAY_DEFINE(_non_z_stacks,
> +                            1,
> +                            CONFIG_ISR_STACK_SIZE);
> ```
> 
> If this is a non Zephyr stack as its name implies, and to be used by non Zephyr code, then this has no business here. You >should allocate stack space alongside the code that will be using it. Furthermore the `K_KERNEL_STACK_ARRAY_DEFINE()` >allocates more than just the provided size so to be compatible with PMP stack guard which non-Zephyr code is not likely to >benefit from.
I used K_KERNEL_STACK_ARRAY_DEFINE to define the stacks as this macro also looks after minimum sizing and alignment requirements, it thought it safer that way...

con-pax commented:
Hi @npitre, 
>     * Please make a separate PR for the `z_smp_cpu_mobile()` changes.
>       This is not RISC-V specific and deserves its own PR.
done
> 
>     * Then you should introduce `CONFIG_RV_BASE_CPU` in a commit of its own.
>       That is on itself pretty uncontrovertial although it has implications
>       in may different places and should stand on its own.
>
done
>     * Please avoid construct like this:
> 

> 
>     * You are reintroducing the memory corruption bug that was fixed in
>       PR [riscv: smp: fix secondary cpus' initial stack #50459](https://github.com/zephyrproject-rtos/zephyr/pull/50459).
> 
done

And we are working on testing the rest.

npitre commented:
> These will only work if the macro is defined and will produce compiler
> errors otherwise hence the #if/#ifdef 

If you use `#if CONFIG_RV_BASE_CPU > 0` then the macro must always 
be defined.

npitre commented:
> The mcounteren register needs to be set on real hardware otherwise
> the registers won't be accessible from user mode... - I don't know 
> if QEMU honors this but i suspect not if the test was passing 
> previously.

Maybe it is a good thing that user mode doesn't have unrestricted access
to those registers? That all depends on the application.

In any case, I'd suggest configuring such things in `arch_kernel_init()`
instead.

petermcs-01 commented:
> 



> 



> > These will only work if the macro is defined and will produce compiler
> > errors otherwise hence the #if/#ifdef
> 
> If you use `#if CONFIG_RV_BASE_CPU > 0` then the macro must always be defined.

No, because the undefined macro resolves to 0...

npitre commented:
> No, because the undefined macro resolves to 0...

I'm pretty sure the MISRA police would disapprove of such usage.

You could move the corresponding Kconfig entry next to the SMP one
as this could potentially be used on any SMP systems and then it would 
always be defined.

fkokosinski commented on arch/riscv/Kconfig:
```diff
@@ -128,6 +128,46 @@ config GEN_IRQ_VECTOR_TABLE
 
+config HAS_MONITOR_HART
+	bool "Does SOC have monitor hart with different capabilities to application harts"
+	default n
```
Booleans default to `n`. This is redundant

con-pax replied:
done

fkokosinski commented on arch/riscv/Kconfig:
```diff
@@ -128,6 +128,46 @@ config GEN_IRQ_VECTOR_TABLE
+
+config MONITOR_HART_SUPERVISOR
+	bool "Monitor hart supports supervisor mode"
+	default n
```
Ditto

con-pax replied:
done

fkokosinski commented on arch/riscv/Kconfig:
```diff
@@ -128,6 +128,46 @@ config GEN_IRQ_VECTOR_TABLE
+
+config SUPERVISOR_MODE_CAPABLE
+	bool "Indicates that supervisor mode is available"
+	default y
```
Why is `CONFIG_SUPERVISOR_MODE_CAPABLE` y-selected by default?

petermcs-01 replied:
Should indeed be n

con-pax replied:
done

fkokosinski commented on arch/riscv/Kconfig:
```diff
@@ -199,6 +246,7 @@ config CMSIS_THREAD_MAX_STACK_SIZE
 config CMSIS_V2_THREAD_MAX_STACK_SIZE
 	default 1024 if 64BIT
 
+
```
Nitpick: unnecessary blank line

con-pax replied:
done

fkokosinski commented on arch/riscv/core/isr.S:
```diff
@@ -60,11 +60,19 @@
+#ifdef CONFIG_RV_BASE_CPU
 #define GET_CURRENT_CPU(dst, tmp) \
 	csrr tmp, mhartid				;\
+	addi tmp, tmp, -CONFIG_RV_BASE_CPU 	;\
```
This begs that `CONFIG_RV_BASE_CPU` defaults to 0 here, since both declaration of the `GET_CURRENT_CPU` macro are basically identical.

fkokosinski commented on arch/riscv/core/reset.S:
```diff
@@ -21,6 +22,8 @@ GTEXT(riscv_cpu_wake_flag)
 GTEXT(riscv_cpu_sp)
 GTEXT(z_riscv_secondary_cpu_init)
 
+GDATA(_non_z_stacks)
```
`non_z`?

petermcs-01 replied:
"Non Zephyr" i.e. stacks for cpus not participating in the zephyr cluster.

fkokosinski replied:
And why do we need this in the general arch code? I mean, this looks to me as something mpfs_icicle-specific, or at least specific to your use case.

Could you describe in more detail (and ideally - add such description to the code itself) why and how you use this?

petermcs-01 replied:
The use case for SMP to date has been rather simplistic with a simple CPU setup that assumes Zephyr has total access to all the CPUs on the system (or at least the first n...). Myself and the team I work with have been involved with low level code development on the MPFS and initially the FU540 for some time now and have seen quite a number of use cases for multi CPU SOCs being looked at by clients. Real world examples have included simple SMP setups, AMP setups involving different combinations of RTOS/Linux/bare metal and other ways to make use of the available cores. With this in mind, I looked at how an RTOS like Zephyr could fit into these type of systems. That's why I settled on the scheme in this code to try and allow the Zephyr component of such systems share the CPU set in a flexible fashion. I also needed to be able to support non-homogeneous cpu setups as MPFS and FU540 are such systems and I am aware of some others on the horizon. Part of the development also looks at operation of Zephyr on real SMP hardware which does not always follow the operation in systems such as QEMU or Renode as they tend to make some assumptions about the low level operation of the CPUs which do not hold true in real life... So, whilst the development has been carried out on MPFS, issues I am dealing with are more generally applicable.

fkokosinski commented on arch/riscv/core/reset.S:
```diff
@@ -38,17 +41,66 @@ SECTION_FUNC(reset, __reset)
+	 * Enable all counters for user mode otherwise some user mode samples
+	 * will fail on real hardware...
+	 */
+	li t0,0xFFFFFFFF
```
Can you use the `BIT_MASK` macro here? Also, your changes could use better indentation

petermcs-01 replied:
I used an explicit value here as although I've decided to enable all the counters, it might be tuned to only include some of them depending on what's really required.

I'm not sure what you mean about the indentation?

fkokosinski replied:
For example, line 93 has a different indentation style than line 99 in the snippet above.

petermcs-01 replied:
Thanks for the clarification, I'll look into it with Conor.

fkokosinski commented on soc/riscv/riscv-privilege/Kconfig:
```diff
@@ -30,4 +30,21 @@ config RISCV_MTVEC_VECTORED_MODE
+config RV_MP_TOTAL_NUM_CPUS
+	int "Total number of CPUs/cores"
+	default 1
+	range 1 5
```
Why was this range chosen? Seems artificial

petermcs-01 replied:
CONFIG_MP_NUM_CPUS is limited to 1 to 4 so I set a range of 1 to 5 for this to cover the cases where a monitor core is in the system. There are some RISC-V chips with even more cores but I had to pick something so I at least covered the system I was testing against...

npitre commented:
> I keep getting load access faults. I can see that I am missing something with
> an offset here: ...

You appear to be missing the `csrr t0,mhartid` from your disassembly.

Also the reported label won't match as you are explicitly subtracting 
an offset from it (which will be compensated by the offset in the 
hartid value).

con-pax commented:
Thanks @npitre,
You were right, I missed reading the mhartid into tmp. It is working on my end now, thanks!

fkokosinski commented on arch/riscv/Kconfig:
```diff
@@ -190,7 +238,14 @@ config MAIN_STACK_SIZE
+	default 4096 if 64BIT
+
 config TEST_EXTRA_STACK_SIZE
+	default 4096 if 64BIT
```
Is this change related to your platform only? If that's the case, please put it in your SoC's `Kconfig.defconfig`.

petermcs-01 replied:
My experience when running some of the tests and sample applications is that the stack sizes are insufficient for Risc-V 64bit especially when compiling for debugging. Even more so when FPU support is added in! There are a number of apps in the Zephyr repository that have fixed stack sizes that I've had to tweak and where possible I've done it via TEST_EXTRA_STACK_SIZE.

fkokosinski commented:
Btw. these changes cause some checks to fail (twister runs and compliance checks). Could you take a look at them?

petermcs-01 commented:
> Btw. these changes cause some checks to fail (twister runs and compliance checks). Could you take a look at them?

Conor is working on them.

npitre commented:
@petermcs-01 commented:

> | config TEST_EXTRA_STACK_SIZE
> |+       default 4096 if 64BIT
> 
> My experience when running some of the tests and sample applications 
> is that the stack sizes are insufficient for Risc-V 64bit especially 
> when compiling for debugging. Even more so when FPU support is added in!

That deserves a separate commit with the above explanation in
the commit log.

con-pax commented:
> > > Btw. these changes cause some checks to fail (twister runs and compliance checks). Could you take a look at them?
> > 
> > 
> > Conor is working on them.
> 
> Hi! Any progress on that front? :)

Hi @fkokosinski, 
Full focus on it this week :)
Kind Regards

fkokosinski commented on include/zephyr/arch/riscv/arch_inlines.h:
```diff
@@ -17,6 +17,9 @@ static ALWAYS_INLINE uint32_t arch_proc_id(void)
 	__asm__ volatile("csrr %0, mhartid" : "=r" (hartid));
+#if CONFIG_RV_BASE_CPU > 0
+		hartid -= CONFIG_RV_BASE_CPU;
+#endif
```
If `CONFIG_RV_BASE_CPU` defaults to `0` anyway, I think we can skip this `if` directive here. It clutters the code a bit, and the compiler will optimize the 0-subtraction anyway.

con-pax replied:
Hi @fkokosinski,
Thanks for pointing this out. done.

edersondisouza reviewed (changes requested):
I'm a bit tired now, we'll keep reviewing next week.

A general comment: maybe this could be simplified from Zephyr point of view if we add something like a "Zephyr CPUs window" or so. We define a range of CPUs that zephyr should be aware. In this window, everything is handled mostly as is now (with the obvious need for a FIRST_CORE or so). If arch code finds a CPU out of this range, it delegates to some SOC code. In the SOC code, you take care of creating "safety stacks" or any other specific stuff. How does that sound? Would it work?

This could save arch code from most of these complexities, and SOC would take care of what is really special to it.

petermcs-01 commented on arch/riscv/core/isr.S:
```diff
@@ -62,7 +62,8 @@
 #ifdef CONFIG_SMP
 #define GET_CURRENT_CPU(dst, tmp) \
 	csrr tmp, mhartid				;\
-	la dst, _kernel + ___kernel_t_cpus_OFFSET	;\
```
From personal experience working with multicore Risc-V systems. The approach used in this code has worked well, particularly when trying to run the system under debuggers where the wfi instruction is likely to see wake unexpected ups. QEMU does not accurately emulate this type of behaviour...

petermcs-01 commented:
> A general comment: maybe this could be simplified from Zephyr point of view if we add something like a "Zephyr CPUs window" or so. We define a range of CPUs that zephyr should be aware. In this window, everything is handled mostly as is now (with the obvious need for a FIRST_CORE or so). If arch code finds a CPU out of this range, it delegates to some SOC code. In the SOC code, you take care of creating "safety stacks" or any other specific stuff. How does that sound? Would it work?
> 
> This could save arch code from most of these complexities, and SOC would take care of what is really special to it.

Our initial approach for this tried to do it at a higher level which would have been more generally applicable but was shot down so we went back to the drawing board and just concentrated on the Risc-V. If it is seen to work well there, I'd like to see it expanded to a more general scheme. For those multi core systems that want to use Zephyr as one part of the solution (Linux + Zephy, Zephyr + bare metal or other mixes) the allocation of cores needs to be addressed.

petermcs-01 commented on arch/riscv/core/reset.S:
```diff
@@ -67,44 +113,120 @@ boot_first_core:
+#if CONFIG_RV_MP_TOTAL_NUM_CPUS > 1
+	j irq_stack_set
+/*
+ * Now need to allocate stack for non zephyr harts for safety.
```
See my comment above, ideally only the cores to run Zephyr should get here but...

petermcs-01 commented on arch/riscv/core/reset.S:
```diff
@@ -38,17 +41,60 @@ SECTION_FUNC(reset, __reset)
+	la sp, _non_z_stacks
+	li t0, __z_interrupt_stack_SIZEOF
+	add sp, sp, t0
+no_monitor_loop:
```
I named it that because it is the loop used when we are not using the monitor hart for Zephyr.

petermcs-01 commented on arch/riscv/core/reset.S:
```diff
@@ -67,44 +113,120 @@ boot_first_core:
 	/* Pre-populate all bytes in z_interrupt_stacks with 0xAA */
 	la t0, z_interrupt_stacks
+#if defined(CONFIG_MP_MAX_NUM_CPUS) && (CONFIG_MP_MAX_NUM_CPUS > 0)
+	li t1, __z_interrupt_stack_SIZEOF * CONFIG_MP_MAX_NUM_CPUS
```
z_interrupt_stacks is an array of stacks and not a single stack. The start up core fills all the zephyr interrupt stacks with AAs, the previous code only did it for the first stack.

edersondisouza replied:
Got it.

petermcs-01 commented on arch/riscv/core/reset.S:
```diff
@@ -67,44 +113,120 @@ boot_first_core:
+#endif
 	add t1, t1, t0
+	/* Only do this once on the startup hart - slightly risky as we are relying
+	 * on other harts not using the stack until we allow them to wake...*/
```
The stack is only initialised by the first core, my comment here is about the fact that one core is writing to all the stacks and we assume that this is the only core that has executed far enough to be using a stack. It's a very small risk and the alternative is to have each core initialise its own stack but this adds more complexity to the start up code.

npitre commented:
> Startup for multiple cores is always tricky, depending on the way the 
> system boots some or all of the cores may execute the Zephyr startup 
> code. In the ideal case, the cores that are dedicated to Zephyr will be 
> the only ones to execute it but if something goes wrong then the 
> startup code may get executed by a core that is not supposed to be 
> running Zephyr. I've tried to accommodate the worst case situation 
> and ensure that if this happens the cores can be safely parked but to do this it is necessary to have a stack 
> for each core.

I don't really understand the argument here. A core is supposed to be 
running Zephyr or it is not. If it is not, then in the worst case situation
where it gets into the Zephyr entry code it will simply stall there looping
forever. No need for a stack to park such core either as you may just park
them using assembly code. No need for fancy parking either as this is
not something that should normally happen in production. Those extra 
stacks represent wasted memory that could be put to better use.

petermcs-01 commented:
> > Startup for multiple cores is always tricky, depending on the way the
> > system boots some or all of the cores may execute the Zephyr startup
> > code. In the ideal case, the cores that are dedicated to Zephyr will be
> > the only ones to execute it but if something goes wrong then the
> > startup code may get executed by a core that is not supposed to be
> > running Zephyr. I've tried to accommodate the worst case situation
> > and ensure that if this happens the cores can be safely parked but to do this it is necessary to have a stack
> > for each core.
> 
> I don't really understand the argument here. A core is supposed to be running Zephyr or it is not. If it is not, then in the worst case situation where it gets into the Zephyr entry code it will simply stall there looping forever. No need for a stack to park such core either as you may just park them using assembly code. No need for fancy parking either as this is not something that should normally happen in production. Those extra stacks represent wasted memory that could be put to better use.

If I had a Euro for every time something unexpected happened in production, I'd have trouble closing my wallet :) I prefer to manage the available CPUs in a conservative fashion as almost 40 years of embedded development has left me with a few scars from making optimistic assumptions. As another couple of examples, many cases with mixed systems on multiple cores, the development is handled by different teams and the code for some of the cores will not be available or an application may be qualified to run on for example 2 cores but supply issues may mean a 4 core device has to be used.
I think there are optimisations that can be made once people have had a chance to use this code but I'm trying not to overload what is already a complex enough set of changes.

petermcs-01 commented on include/zephyr/arch/riscv/arch_inlines.h:
```diff
@@ -17,6 +17,7 @@ static ALWAYS_INLINE uint32_t arch_proc_id(void)
 
 #ifdef CONFIG_SMP
 	__asm__ volatile("csrr %0, mhartid" : "=r" (hartid));
+	hartid -= CONFIG_RV_BASE_CPU;
```
Within a given implementation the rules for mapping between the real cpu ID and the zephyr cpu ID are generally known and I would expect that anywhere that the value is used this would be catered for.

con-pax replied:
Hi @andyross, 
Thanks for this suggestion, I am starting to look into this. I will try include it with our next rebase

con-pax replied:
Hi @andyross 
I have implemented two API's locally (one to get zephyr cpu id from proc_id and vice versa) and have tested them. I have put the prototypes in zephyr/include/zephyr/sys/arch_interface.h:
```
/**
 * @brief Returns Zephyr CPU ID when arch-specific CPU ID's differ.
 * 
 * In multiprocessor architectures, the low-level unique ID value
 * associated with the current CPU is not guaranteed to match
 * Zephyr's own sequential CPU ID's. This function allows the 
 * translation between an architectures CPU ID to a Zephyr
 * CPU ID.
 *
 * @return Zephyr CPU ID translated from arch-specific CPU ID.
*/

static inline uint32_t arch_proc_id_to_cpu(uint32_t proc_id);

/**
 * @brief Returns arch-specific CPU ID, given a Zephyr CPU ID.
 *
 * The inverse of arch_proc_id_to_cpu(uint32_t proc_id). This
 * function gets the architecture-specific unique CPU ID
 * from a given Zephyr CPU ID.
 *
 * @return Architecture-specific unique CPU ID/
*/

static inline uint32_t arch_cpu_to_proc_id(uint32_t cpu_id);
```
and have defined the above in zephyr/include/zephyr/arch/riscv/arch_inlines.h:
```
static ALWAYS_INLINE uint32_t arch_proc_id_to_cpu(uint32_t hartid)
{
	uint32_t cpu_id;
	cpu_id = hartid -= CONFIG_RV_BASE_CPU;
	return cpu_id;
}

static ALWAYS_INLINE uint32_t arch_cpu_to_proc_id(uint32_t cpu_id)
{
	uint32_t hartid;
	hartid = cpu_id + CONFIG_RV_BASE_CPU;
	return hartid;
}
```
I want to ask you if you think this is suitable? And also if you think it is, should I open a separate Pull Request, just for the above?
Would really appreciate any guidance on this. 
Cheers!

petermcs-01 commented on arch/riscv/core/reset.S:
```diff
@@ -38,17 +41,60 @@ SECTION_FUNC(reset, __reset)
+	sr a1, 0(a2)
+#if CONFIG_HAS_MONITOR_HART
+/*
+ * For SMP operation the monitor hart cannot be part of the SMP group
```
My rational for making a special case for the monitor core is that in many instances those types of cores have a very different feature set to the primary applications cores and you therefore have to be very careful about what code you let them execute. It muddies the water somewhat but at some level the asymmetric nature of some of the Risc-V multi-core SOCs needs to be handled.

npitre commented:
I'm not totally green either when it comes to embedded system experience 
so let's sstick to technical arguments please.

Again, for each CPU you have 2 choices:

1) you want Zephyr to take care of it, or

2) you don't want Zephyr to take care of it.

In case of (1) you allocate a stack for it and let it run when flagged
to run past the stopgate. Simple.

In case of (2) you stall it in an infinite loop in the early assembly 
code. It doesn't have to be more sophisticated than that. That's what
the current code does already in fact. No need for a stack. This isn't 
supposed to happen anyway because, if Zephyr doesn't take care of this 
CPU, something else should, and the Zephyr code should never be reached
by such CPU in the first place. Here we simply make sure the stopgate stays
in place in case things aren't right until they're fixed.

In other words, let's make things _simpler_ not more complex.

andyross reviewed (commented):
I've been mostly nodding along with this, but should comment I guess.  On the whole it looks broadly OK to me; it's all in the arch layer and isn't able to confuse the kernel internals, nor force them to evolve in weird ways.

Just as general note: I do tend to agree with @npitre that the core initialization handling here seems kinda convoluted.  Zephyr permits a **lot** of latitude for how an arch initializes its CPUs (e.g. `arch_start_cpu()` is an async thing that just needs a callback on the interrupt stack at some point in the finite future), so it's a little weird to see all this code handling all these circumstances.  There just shouldn't be that many circumstances internal to one arch!  And if there are, then maybe that argues you should have some kind of internal platform API so the SOC code can manage startup itself instead of adding all the #if'ery here?

con-pax commented:
> I'm not totally green either when it comes to embedded system experience so let's sstick to technical arguments please.
> 
> Again, for each CPU you have 2 choices:
> 
> 1. you want Zephyr to take care of it, or
> 2. you don't want Zephyr to take care of it.
> 
> In case of (1) you allocate a stack for it and let it run when flagged to run past the stopgate. Simple.
> 
> In case of (2) you stall it in an infinite loop in the early assembly code. It doesn't have to be more sophisticated than that. That's what the current code does already in fact. No need for a stack. This isn't supposed to happen anyway because, if Zephyr doesn't take care of this CPU, something else should, and the Zephyr code should never be reached by such CPU in the first place. Here we simply make sure the stopgate stays in place in case things aren't right until they're fixed.
> 
> In other words, let's make things _simpler_ not more complex.

Hi @npitre,
We are finishing up testing the code with the removal of the stack creation for non zephyr related harts, It will be in our next rebase.

npitre commented:
@andyross wrote:

> I tend to agree that all the magic offsetting is going to be confusing
> and fragile.  (e.g., what happens when the list of Zephyr CPUs becomes
> non-contiguous in the "riscv" ID space?  You know that's going to happen
> at some point)

The solution is, of course, to dispense with the conversion entirely.
We implemented a similar scheme for ARM64 already.

I created PR #53572 that allows for any CPU to be assigned to Zephyr in
any order and regardless of their actual hardware ID.

@con-pax wrote:

> I have implemented two API's locally (one to get zephyr cpu id from 
> proc_id and vice versa) and have tested them. I have put the prototypes
> in zephyr/include/zephyr/sys/arch_interface.h
> 
> I want to ask you if you think this is suitable?

In my opinion it is not. This will always be suboptimal. The most 
important reason for converting the hartid to the Zephyr CPU number is
about retrieving the `_current_cpu` structure pointer. So why not having
each CPU hold that pointer directly instead?

Please review PR #53572. That will solve all your hartid related issues
in a much more elegant way.

con-pax commented:
> 
> The solution is, of course, to dispense with the conversion entirely. We implemented a similar scheme for ARM64 already.
> 
> I created PR #53572 that allows for any CPU to be assigned to Zephyr in any order and regardless of their actual hardware ID.
> 
> @con-pax wrote:
> 
> > I have implemented two API's locally (one to get zephyr cpu id from
> > proc_id and vice versa) and have tested them. I have put the prototypes
> > in zephyr/include/zephyr/sys/arch_interface.h
> > I want to ask you if you think this is suitable?
> 
> In my opinion it is not. This will always be suboptimal. The most important reason for converting the hartid to the Zephyr CPU number is about retrieving the `_current_cpu` structure pointer. So why not having each CPU hold that pointer directly instead?
> 
> Please review PR #53572. That will solve all your hartid related issues in a much more elegant way.

Hi @npitre ,

Thanks, this looks really cool! I can test it out on Monday. I do have a comment, which I will leave in review.
Cheers!

con-pax committed:
riscv: enable booting from non-zero indexed RISC-V hart

RISC-V multi-hart systems that employ a heterogeneous core complex are
not guaranteed to have the smp capable harts starting with a unique id
of zero, matching Zephyr's sequential zero indexed cpu numbering scheme.

Add option, RV_BOOT_HART to choose the hart to boot from.
On reset, check the current hart equals RV_BOOT_HART: if so, boot first
core. else, loop in the boot secondary core and wait to be brought up.

For multi-hart systems that are not running a Zephyr mp or smp
application, park the non zephyr related harts in a wfi loop.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

con-pax commented:
Hi @npitre , @edersondisouza
I have refactored our SMP code and pushed. Now here comes the problem: 
The CI seems to be acting erratic and Qemu builds are failing for reasons not known to me (maybe hitting resource limits?).
Test that passed on the previous run fail on this run, against the same code.

I, like most, hate hearing this phrase: but the failing builds in the CI work locally on my machine, which I can replicate.
As an example: I will attach the output of the ```zephyr/tests/kernel/sched/schedule_api/``` job which failed in the twister build, but passed on my machine. (this job passed in the CI in the previous run).
Here is the link to the CI: https://github.com/zephyrproject-rtos/zephyr/actions/runs/3967255730/jobs/6799021347 

Attached is the output of running on my local machine.
[output.log](https://github.com/zephyrproject-rtos/zephyr/files/10466740/output.log)

Unsure on what to do here: do I just keep rebasing and force pushing and hope there is a time that the CI is not under load?
Cheers.

npitre commented:
Sorry for this late review. But I think a few details still deserve
to be clarified.

First, please make a separate PR for the timer fix. This one is a
no-brainer and will be merged immediately. Just consider making the
subject line a bit more precise e.g. "riscv: timer: ..."

Next, let's reconsider this `CONFIG_RV_BASE_CPU`. I personally don't like
it much as it still implies a linear CPU range. I think this should be
`CONFIG_RV_BOOT_HART` with a default of 0 (it could eventually be -1
for "don't care"). This way you unambiguously determine which CPU you
want to be the first one to boot Zephyr, and all the others will be
discovered from DT.

The reset.S won't have a clear range to test as the boot CPU might not
necessarily be the first CPU in the range, but you can drop that change
for now. Just let unexpected CPUS wait forever in the boot_secondary_core
loop (they should normally not get there in the first place).

Then you may combine both the reset.S and the Kconfig changes in the same
commit as they are closely related.

Next, when enumerating hartids for each CPU, you need to be careful about
assigning the proper hartid to CPU 0 and skipping that hartid afterwards.
To make the code cleaner, that enumeration should happen whether or not
CONFIG_SMP is set (as long as CONFIG_MP_MAX_NUM_CPUS > 1). So making
the hartid variable in `struct _cpu_arch` could be dependent on:

```
#if defined(CONFIG_SMP) || (CONFIG_MP_MAX_NUM_CPUS > 1)
```

rather than being available all the time. And again, those 2 commits
could also be combined as they're closely related.

con-pax commented:
> Sorry for this late review. But I think a few details still deserve to be clarified.

@npitre, thanks a million! I will get on top of these tomorrow.
Cheers

con-pax committed:
riscv: smp: use devicetree to map hartids to zephyr cpus

For RISC-V, the reg property of a cpu node in the devicetree describes
the low level unique ID of each hart. Using devicetree macro's, a list
of all cpus with status "okay" can be generated.

Using devicetree overlays, a hart or multiple harts can be marked as
"disabled", thus excluding them from the list. This allows platforms
that have non-zero indexed SMP capable harts to be functionally mapped
to Zephyr's sequential CPU numbering scheme.

On kernel init, if the application has MP_MAX_NUM_CPUS greater than 1,
generate the list of cpu nodes from the device tree with status "okay"
and  map the unique hartid's to zephyr cpu's

While we are at it, as the hartid is the value that gets passed to
z_riscv_secondary_cpu_init, use that as the variable name instead of
cpu_num

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

fkokosinski reviewed (changes requested):
Hi @con-pax, it looks like after your recent push resulted in the test in `tests/kernel/common` failing on the `hifive_unleashed` platform.

I've checked it locally and this test works as expected in Renode simulation, but it takes some time to pass. This test uses a boot delay of 500ms, so if your recent push somehow affected negatively boot time on multicore RISC-V platforms, then this could be the cause of this behavior. Could you look into it? I think we could use an extra config for RISC-V platforms for this test, if this is behavior is the same on other multicore RISC-V platforms.

con-pax commented:
> Hi @con-pax, it looks like after your recent push resulted in the test in `tests/kernel/common` failing on the `hifive_unleashed` platform.
> 
> I've checked it locally and this test works as expected in Renode simulation, but it takes some time to pass. This test uses a boot delay of 500ms, so if your recent push somehow affected negatively boot time on multicore RISC-V platforms, then this could be the cause of this behavior. Could you look into it? I think we could use an extra config for RISC-V platforms for this test, if this is behavior is the same on other multicore RISC-V platforms.

@fkokosinski, 
I will start testing this now.
Cheers!

akabaev commented on arch/riscv/include/kernel_arch_func.h:
```diff
@@ -36,6 +36,20 @@ static ALWAYS_INLINE void arch_kernel_init(void)
+	unsigned int cpu_num, hart_x;
+
+	for (cpu_num = 1, hart_x = 0; cpu_num < CONFIG_MP_MAX_NUM_CPUS; cpu_num++) {
+		if (cpu_node_list[hart_x] == CONFIG_RV_BOOT_HART) {
```
Isn't the point of this code really to skip the currently running 'boot' cpu (a.k.a cpu running this code), which has been setup already above and is guaranteed to have hartid  of  CONFIG_RV_BOOT_HART? Comparing with _kernel.cpus[0].arch.hartid value read from csr_read(mhartid); will spell it more explicitly.
Should the same trick as with cpu_node_list be used to adjust the value returned by arch_num_cpus(). The only other architecture that sometimes returns anything, but CONFIG_MP_MAX_NUM_CPUS is xtensa though.

con-pax replied:
>  Comparing with _kernel.cpus[0].arch.hartid value read from csr_read(mhartid); will spell it more explicitly. Should the same trick as with cpu_node_list be used to adjust the value returned by arch_num_cpus(). The only other architecture that sometimes returns anything, but CONFIG_MP_MAX_NUM_CPUS is xtensa though.

@akabaev, Good spot! pushing a change for this shortly.

>  While there, is running _without_ dts is still a thing supported by Zephyr?

I certainly dont think the riscv_machine_timer would work without dts, as the MTIME register is retrieved from it.

akabaev replied:
@con-pax This was not really a request, just some musings inspired by looking at your changes. For one, I do not have whole lot of the background on the while CONFIG_MP_NUM_CPUS -> CONFIG_MP_MAX_NUM_CPUS and their relationship to  arch_num_cpus() and whether adjustment of  arch_num_cpus() based on DTS content is even desired. My question about running without DTS was more about this: should you be providing fallback cpu_id == mhartid path when there is no DTS to build cpu_node_list from?

con-pax replied:
I guess the point of this PR is to be able to run applications on platforms that have hart id's that are not aligned with Zephyrs sequential cpu numbering scheme. 
The proposed changes here allow for both non-zero indexed and non-contiguous SMP capable hart configurations, thanks to being able to parse the device tree. 
As I see it, the right way to do this is to use the Devicetree. Other implementations may involve more Kconfig options and/or may only allow you to use a contiguous range of hartids. 
The inspiration for this change set was borrowed from what is done in arm64 where they use the same devicetree macros to define a cpu_node_list, which would also fall over if devicetree support was not there.

con-pax commented:
> Hi @con-pax, it looks like after your recent push resulted in the test in `tests/kernel/common` failing on the `hifive_unleashed` platform.
> 
> I've checked it locally and this test works as expected in Renode simulation, but it takes some time to pass. This test uses a boot delay of 500ms, so if your recent push somehow affected negatively boot time on multicore RISC-V platforms, then this could be the cause of this behavior. Could you look into it? I think we could use an extra config for RISC-V platforms for this test, if this is behavior is the same on other multicore RISC-V platforms.

Hi @fkokosinski, I found the issue and implemented a fix for it. The issue is that originally, out of reset, a hartid was checked and if it was the chosen boot hart it would jump to boot first core loop, else it would jump to boot secondary core loop. However, for Multi-hart systems that are not running an MP or SMP application (i.e CONFIG_MP_MAX_NUM_CPUS =1) the non zephyr harts would be spinning around in the boot secondary core loop. The fix is to park non-zephyr related cores into a wfi loop. I have tested this on our platform, as well as renode and works as expected

edersondisouza reviewed (approved):
It seems the PR title and content diverged a bit, other than that, LGTM.
